src/librbd/cache/pwl/rwl/WriteLog.cc

```diff
@@ -552,6 +556,10 @@ bool WriteLog<I>::retire_entries(const unsigned long int frees_per_tx) {
       ceph_assert(this->m_first_valid_entry == initial_first_valid_entry);
       this->m_first_valid_entry = first_valid_entry;
       this->m_free_log_entries += retiring_entries.size();
+      if (m_cache_state->empty == false && m_first_free_entry == m_first_valid_entry) {
```

Comment:
```suggestion
      if (!m_cache_state->empty && m_first_free_entry == m_first_valid_entry) {
```

Reply:
Also, why is the true -> false transition in `alloc_op_log_entries` based on `m_log_entries` emptiness while the false -> true transition here based on comparing head and tail pointers?  I realize that the head pointer is expected to be equal to the tail pointer only when the cache is empty but I don't understand the reason for this asymmetry.

Reply:
ok, then change all using `m_first_free_entry == m_first_valid_entry`. 

I used to call `m_log_entries.empty()`, but there was a problem here(`retire_entries`),  `m_log_entries` and `m_first_valid_entry` are not changed in a code block, I can only update it after `m_first_valid_entry` is modified, but then `m_log_entries` may have been change because `m_lock` don't lock two code blocks.